File: BC-W4/PHP/PenandPaper/Pen.php
<?php
    class Pen {
        private $inkCapacity;
        private $inkAmount;
        
        private function validate($value) {
            if ( is_numeric($value) ) {
                return $value;
            }
            throw new Exception("Invalid param!");
        }
        
        public function __construct($inkCapacity = 4096) {
            $this->inkCapacity = $this->validate($inkCapacity);
            $this->inkAmount = $this->inkCapacity;
        }
        
        public function __get($name) {
            if ( property_exists( $this, $name) ) {
                return $this->$name;
            }
            throw new Exception("Attribute error: attribute $name not found");
        }
        
        public function write(Paper $paper, string $message) {
            $this->inkAmount -= strlen($message);
            
            if ( $this->inkAmount < 0 ) {
                throw new Exception("Out of Ink");
            }
            
            $paper->addContent($message);
            
        }
        
        public function refill() {
            $this->inkAmount = $this->inkCapacity;
        }
        
    }
?>

File: BC-W3/TaskCpp/bcw4/Pen&Paper/Pen.cpp
#include "Pen.h"

Pen::Pen(int inkCapacity) {
    this->inkCapacity = inkCapacity;
    this->inkAmount = inkCapacity;
}
Pen::~Pen() {}

int Pen::getInkAmount() const {
    return this->inkAmount;
}
int Pen::getInkCapacity() const {
    return this->inkCapacity;
}

void Pen::write(Paper& paper, const std::string& message) {
    this->inkAmount -= message.size();
    
    if ( this->inkAmount < 0 ) {
        throw OutOfInk();
    }
    
    paper.addContent(message);
}
void Pen::refill() {
    this->inkAmount = this->inkCapacity;
}